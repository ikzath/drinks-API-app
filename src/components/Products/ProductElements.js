import styled from 'styled-components';
import {BiDetail, BiDish} from 'react-icons/bi';
import {Container} from '../../GlobalStyles';

export const ProductContent = styled.section`
    display: flex;
    justify-content: center;
    flex-direction: column;
    align-items: center;
    height: 90rem;
`;

export const ProductContainer = styled(Container)`
    display: grid;
    grid-template-columns: repeat(3, 1fr);
    grid-auto-rows: 1fr;
    grid-gap: 20px;
    align-items: left;
    padding: 5rem 8rem;
    color: #333333;
    ${Container};
`;

export const ProductTitle = styled.h2`
    font-size: 4rem;
    font-weight: bold;
    color: orange;
    `;

    export const ProductCardContent = styled.div`
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-top: 5rem;
    @media only screen and (max-width:900px){
        flex-direction: column;
    }
`;

export const ProductCard = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;
    width: 30rem;
    height: 30rem;
    background-color: #fff;
    box-shadow: 0px 5px 40px rgba(0, 0, 0, 0.19);
    border-radius: 20px;
    transition: all .5s ease;
&:not(:last-child) {
    margin-right: 5rem;
    @media only screen and (min-width: 1300px){
        margin-right: 10rem;
    }
    @media only screen and (min-width: 1500px){
        margin-right: 20rem;
    }
    @media only screen and (max-width:900px){
        margin-bottom: 10rem;
    }
    @media only screen and (max-width: 800px){
        box-shadow: 0px 5px 20px rgba(0, 0, 0, 0.12);
    }
}
@media only screen and (max-width:900px){
        width: 20rem;
        margin-right: 5rem;

}
@media only screen and (max-width:500px){
        width: 20rem;
        margin-right: 5rem;
}
@media only screen and (min-width:1890px){
        width: 20rem;
        margin-right: 5rem;
}
&:hover {
    box-shadow:  0px 10px 80px rgba(0, 0, 0, 0.21);
    transform: scale(1.05);
    background-color: #E38B06;
    color: #fff;
    
}
`;

export const ProductIconContainer = styled.div`
    width: 12rem;
    height: 12rem;
    transition: all .3s ease-out;
    margin-top: 2rem;
    display: flex;
    justify-content: center;
    align-items: center;
`;
export const ProductIcon = styled.img`
    width: 100px;
    height: 100px
`;

export const ProductIcon2 = styled(BiDetail)`
    color: #333;
    font-size: 6rem;
`;
export const ProductIcon3 = styled(BiDish)`
    color: #333;
    font-size: 6rem;
`;

export const ProductCardTitle = styled.h3`
    font-size: 2rem;
    font-weight: bold;
    padding-top: 1rem;    
`;

export const ProductCardText = styled.p`
    font-size: 2rem;
    padding: 1rem 2rem;
    color: red;
`;

export const Select = styled.select`
    width: 25%;
    height: 60px;
    background: #e6e4dc;
    color: black;
    padding-left: 5px;
    font-size: 14px;
    font-weight: bold;
    border: none;
    margin-left: 10px;
    margin-top: 20px;
    border-radius: 10px;

  option {
    color: black;
    background: white;
    display: flex;
    white-space: pre;
    min-height: 20px;
    padding: 0px 2px 1px;
    background-color: whitesmoke;
  }
`;

export const ButtonContainer = styled.div`
    display: flex;
    margin-right: 20px;
    `

export const ButtonWrapper = styled.section`
    background-color: whitesmoke;
    padding: 20px;
    display: flex;
    justify-content: space-between;
    align-items: center;
`;

export const Button = styled.button`
    border-radius: 20px;
    background-color: ${({primary}) => primary? '#E38B06': 'lightgreen'};
    color: ${({primary}) => primary ? 'whitesmoke': '#fff'};
    padding: 18px 30px;
    font-size: 20px;
    outline: none;
    cursor: pointer;
    border: none;
    transition: transform .2s ease;
    &:hover{
        background-color: ${({primary}) => primary? 'white': '#E38B06'};  
    transform: translateY(-.5rem) scale(1.02);
    color: #000;
    }
`;

export const ProductNoListing = styled.div`
    height: 90rem;
    font-size: 3rem;
`