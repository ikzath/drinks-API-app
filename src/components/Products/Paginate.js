import React from 'react';
import { PaginateButton } from './ProductElements';


const Paginate = ({ postsPerPage, totalPosts, paginate }) => {
  const pageNumbers = [];

  for (let i = 1; i <= Math.ceil(totalPosts / postsPerPage); i++) {
    pageNumbers.push(i);
  }

  return (
    <nav>
      <div style={{ display: 'flex', justifyContent: 'space-evenly' }}>
        {pageNumbers.map(number => (
          <div key={number} className='page-item'>
              <PaginateButton onClick={() => paginate(number)} >{number}</PaginateButton>
          </div>
        ))}
      </div>
    </nav>
  );
};

export default Paginate;