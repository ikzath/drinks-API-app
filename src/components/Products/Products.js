import dataSet from '../../dataSet.json';
import React, { useState, useEffect } from 'react';
import ProductList from '../ProductList/ProductList';
import { Select, Button, ButtonWrapper, ButtonContainer, ProductNoListing } from '../Products/ProductElements';
import Paginate from "../Paginate/Paginate";

const Products = () => {

    const [products] = useState(dataSet);
    const [filterByType, setFilterByType] = useState("All");
    const [filterByBrand, setFilterByBrand] = useState("All");
    const [results, setResults] = useState([]);
    const [currentPage, setCurrentPage] = useState(1);
    const [postsPerPage] = useState(6);

    const getUnique = ((items, value)=> {
        return [...new Set(items.map(item => item[value]))]
    })

    // get unique option value keys
    let types = getUnique(dataSet, 'type')
        types = ['All', ...types]
        types= types.map((option, idx) => (
        <option key={idx} value={option.type}>{option}</option>
    ))

    let brand = getUnique(dataSet, 'brand')
        brand = ['All', ...brand]
        brand= brand.map((option, idx) => (
        <option key={idx} value={option.brand}>{option}</option>
    ))

        function onChangeBrand(e){
            setFilterByBrand(e.target.value)
        }

        function onChangeType(e){
            setFilterByType(e.target.value)
        }

    
    useEffect(() => {
        let result = [...products]; 

        if(filterByType === 'All' && filterByBrand === 'All' ){
            setResults(result)
        }           
        if(filterByType !== 'All') {
            result = result.filter(item => item.type === filterByType);      
        }         
        if(filterByBrand !== 'All') {
            result = result.filter(item => item.brand === filterByBrand);      
        }  
        
        setResults(result)
        }, [ products, filterByBrand, filterByType])

        
        function onClickSort(){
            const dataArray = [ ...results]
            setResults(dataArray.sort((a,b) => a.name.localeCompare(b.name)));
        }

        function onClickSortDesc(){
            const dataArray = [ ...results]
            setResults(dataArray.sort((a,b) => b.name.localeCompare(a.name)));
        }

        function filterByPrice(){
            const dataArray = [ ...results]
            setResults(dataArray.sort((a,b) => a.price >b.price ? 1 : -1));
        }
          
        // Get current posts
        const indexOfLastPost = currentPage * postsPerPage;
        const indexOfFirstPost = indexOfLastPost - postsPerPage;
        const currentPosts = results.slice(indexOfFirstPost, indexOfLastPost);

        // Change page
        const paginate = pageNumber => setCurrentPage(pageNumber);
        
    return (
        <div data-testid='product-1'>
             <label> Filter by type : </label>
            <Select name='products' onChange={onChangeType}>
                {types}
            </Select><br />
            <label> Filter by brand : </label>
            <Select name='products' onChange={onChangeBrand} >
                {brand}
            </Select>
            <div>
            <ButtonContainer>
                <ButtonWrapper>
                    <Button data-testid='button-1' primary onClick={onClickSort}> Sort by ascending order </Button>  
                </ButtonWrapper>
                <ButtonWrapper>
                    <Button data-testid='button-1' primary onClick={onClickSortDesc}> Sort by descending order </Button>  
                </ButtonWrapper>
                <ButtonWrapper>
                    <Button data-testid='button-1' primary onClick={filterByPrice}>Lowest to Highest Price </Button>  
                </ButtonWrapper>
            </ButtonContainer>
            </div>
            {results.length ?
           <ProductList products={currentPosts}/> :  <ProductNoListing>No such perfumes available</ProductNoListing> }
            <Paginate
                postsPerPage={postsPerPage}
                totalPosts={results.length}
                paginate={paginate}
            />        
        </div>
    )
}

export default Products;







