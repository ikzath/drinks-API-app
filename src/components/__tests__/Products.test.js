import { render, screen, cleanup } from "@testing-library/react";
import Products from '../Products/Products';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

afterEach(cleanup);

it('should render products component', ()=>{
    render(<Products />);
    const productElement = screen.getByTestId('product-1');
    expect(productElement).toBeInTheDocument();
})


it('renders without crashing', ()=>{
    const div = document.createElement('div');
    ReactDOM.render(<Products />, div);
})


it('renders button correctly', ()=>{
    render(<Products />);
    const productButton = screen.getByTestId('product-1');
    expect(productButton).toBeInTheDocument();
})

it('matches snapshot', ()=> {
    const tree  = renderer.create(<Products products='all' ></Products>).toJSON();
    expect(tree).toMatchSnapshot();
})