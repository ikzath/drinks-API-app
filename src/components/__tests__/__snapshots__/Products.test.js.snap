// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  data-testid="product-1"
>
  <label>
     Filter by type : 
  </label>
  <select
    className="sc-dIsUp iUEUFW"
    name="products"
    onChange={[Function]}
  >
    <option>
      All
    </option>
    <option>
      Eau de Parfum
    </option>
    <option>
      Eau de Toilette
    </option>
    <option>
      Parfum
    </option>
  </select>
  <br />
  <label>
     Filter by brand : 
  </label>
  <select
    className="sc-dIsUp iUEUFW"
    name="products"
    onChange={[Function]}
  >
    <option>
      All
    </option>
    <option>
      Lancôme
    </option>
    <option>
      Jean Paul Gaultier
    </option>
    <option>
      Paco Rabanne
    </option>
    <option>
      Hugo Boss
    </option>
    <option>
      Giorgio Armani
    </option>
    <option>
      Yves Saint Laurent
    </option>
    <option>
      Chloé
    </option>
    <option>
      Dior
    </option>
    <option>
      Guerlain
    </option>
    <option>
      Narciso Rodriguez
    </option>
    <option>
      Issey Miyake
    </option>
    <option>
      Dolce & Gabbana
    </option>
    <option>
      Lacoste
    </option>
    <option>
      Viktor & Rolf
    </option>
    <option>
      Emporio Armani
    </option>
  </select>
  <div>
    <section
      className="sc-bqGGPW gHHmpt"
    >
      <button
        className="sc-ksluID guRrvX"
        data-testid="button-1"
        onClick={[Function]}
      >
         Sort by ascending order 
      </button>
    </section>
  </div>
  <div>
    No such perfumes available
  </div>
  <div
    className="sc-hBMUJo dGMzYZ"
  />
</div>
`;
